File: Sort/Convert an array to reduced form.java
//https://practice.geeksforgeeks.org/problems/convert-an-array-to-reduced-form/0
//Time complexity : O(nlogn) for sorting O(n^2) in iteration


import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
	public static void main (String[] args) {
		//code
		int t ;
		Scanner sc = new Scanner(System.in);
		t = sc.nextInt();
		while(t-->0)
		{
		    int n;
		    n = sc.nextInt();
		    int[] a = new int[n];
		    int[] b = new int[n];
		    for(int i=0;i<n;i++)
		    {
		        a[i] = sc.nextInt();
		        b[i] = a[i];
		    }
		    Arrays.sort(b);
		    for(int i=0;i<n;i++)
		    {
		        for(int j=0;j<n;j++)
		        {
		            if(a[i]==b[j])
		            {
		                System.out.print(j+" ");
		                break;
		            }
		        }
		    }
		    System.out.print("\n");
 		}
	}
}

File: Hash/Convert an array to reduced form.cpp
//https://www.geeksforgeeks.org/convert-an-array-to-reduced-form-set-1-simple-and-hashing/
//practice.geeksforgeeks.org/problems/convert-an-array-to-reduced-form/0
//Time complexity : O(nlogn) for sorting O(n) for iteration


// C++ program to convert an array in reduced 
// form 
#include <bits/stdc++.h> 
using namespace std; 
  
void convert(int arr[], int n) 
{ 
    // Create a temp array and copy contents 
    // of arr[] to temp 
    int temp[n]; 
    memcpy(temp, arr, n*sizeof(int)); 
  
    // Sort temp array 
    sort(temp, temp + n); 
  
    // Create a hash table. Refer  
    // http://tinyurl.com/zp5wgef  
    unordered_map<int, int> umap; 
  
    // One by one insert elements of sorted 
    // temp[] and assign them values from 0 
    // to n-1 
    int val = 0; 
    for (int i = 0; i < n; i++) 
        umap[temp[i]] = val++; 
  
    // Convert array by taking positions from 
    // umap 
    for (int i = 0; i < n; i++) 
        arr[i] = umap[arr[i]]; 
} 
  
void printArr(int arr[], int n) 
{ 
    for (int i=0; i<n; i++) 
        cout << arr[i] << " "; 
} 
  
// Driver program to test above method 
int main() 
{ 
    int arr[] = {10, 20, 15, 12, 11, 50}; 
    int n = sizeof(arr)/sizeof(arr[0]); 
  
    cout << "Given Array is \n"; 
    printArr(arr, n); 
  
    convert(arr , n); 
  
    cout << "\n\nConverted Array is \n"; 
    printArr(arr, n); 
  
    return 0; 
} 